### ISSUE TITLE
Nothing can be predicted after changing the code

### ISSUE THREAD

**Issue by wafaer** [wang-xinyu/tensorrtx#1413]
## Env

- GPU, 3060
- OS, 20.04
- Cuda 11.4
- TensorRT version 8.2.4.2

## About this repo

- which branch/tag/commit are you using?
- which model? yolov5, retinaface? 
yolov5

## Your problem

Hello, I added the attention module after cv1 in the C3 module of the backbone network. No instances were predicted when generating engine inference. The Python version is normal. What is the reason for this?

##python
class h_sigmoid(nn.Module):
    def __init__(self, inplace=True):
        super(h_sigmoid, self).__init__()
        self.relu = nn.ReLU6(inplace=inplace)

    def forward(self, x):
        return self.relu(x + 3) / 6

class h_sigmoid1(nn.Module):
    def __init__(self, inplace=True):
        super(h_sigmoid1, self).__init__()
        self.relu = nn.ReLU(inplace=inplace)

    def forward(self, x):
        return self.relu(x)

class h_swish(nn.Module):
    def __init__(self, inplace=True):
        super(h_swish, self).__init__()
        self.sigmoid = h_sigmoid1(inplace=inplace)

    def forward(self, x):
        return x * self.sigmoid(x)

class CoordAtt(nn.Module):
    def __init__(self, inp, oup, reduction=32):
        super(CoordAtt, self).__init__()
        self.pool_h = nn.AdaptiveAvgPool2d((None, 1))
        self.pool_w = nn.AdaptiveAvgPool2d((1, None))

        mip = max(8, inp // reduction)

        self.conv1 = nn.Conv2d(inp, mip, kernel_size=1, stride=1, padding=0)
        self.bn1 = nn.BatchNorm2d(mip)
        self.act = h_swish()

        self.conv_h = nn.Conv2d(mip, oup, kernel_size=1, stride=1, padding=0)
        self.conv_w = nn.Conv2d(mip, oup, kernel_size=1, stride=1, padding=0)

    def forward(self, x):
        identity = x

        n, c, h, w = x.size()
        x_h = self.pool_h(x)

        x_w = self.pool_w(x).permute(0, 1, 3, 2)

        y = torch.cat([x_h, x_w], dim=2)

        y = self.conv1(y)
        y = self.bn1(y)
        y = self.act(y)

        x_h, x_w = torch.split(y, [h, w], dim=2)

        x_w = x_w.permute(0, 1, 3, 2)

        a_h = self.conv_h(x_h).sigmoid()

        a_w = self.conv_w(x_w).sigmoid()

        out = identity * a_w * a_h

        return out
        
        
##tensorrt
static ILayer* CoordAtt(INetworkDefinition *network, std::map<std::string, Weights>& weightMap, ITensor& input, int& inp, std::string lname)
{
    Dims dims = input.getDimensions();
    int h_size = dims.d[1];
    int w_size = dims.d[2];

    /// pool_h *************************************************************************************************************
    auto h_pool = network->addPoolingNd(input, PoolingType::kAVERAGE,DimsHW{1,1});
    h_pool->setStrideNd(DimsHW{1,1});
    h_pool->setPaddingNd(DimsHW{0,0});
    h_pool->setWindowSizeNd(DimsHW{1,w_size});
    ITensor* h_value = h_pool->getOutput(0);

    Permutation permutation = {1,0,2};
    nvinfer1::IShuffleLayer* shuffleLayer = network->addShuffle(*h_value);
    shuffleLayer->setFirstTranspose(Permutation(permutation));
    ITensor* perm1 = shuffleLayer->getOutput(0);

    /// pool_w *************************************************************************************************************
    auto w_pool = network->addPoolingNd(input, PoolingType::kAVERAGE,DimsHW{1,1});
    w_pool->setStrideNd(DimsHW{1,1});
    w_pool->setPaddingNd(DimsHW{0,0});
    w_pool->setWindowSizeNd(DimsHW{h_size,1});
    ITensor* w_value = w_pool->getOutput(0);

    Permutation permutation1 = {0,2,1};
    nvinfer1::IShuffleLayer* shuffleLayer1 = network->addShuffle(*w_value);
    shuffleLayer1->setFirstTranspose(Permutation(permutation1));
    ITensor* perm2 = shuffleLayer1->getOutput(0);

    Permutation permutation2 = {1,0,2};
    nvinfer1::IShuffleLayer* shuffleLayer2 = network->addShuffle(*perm2);
    shuffleLayer2->setFirstTranspose(Permutation(permutation2));
    ITensor* perm3 = shuffleLayer2->getOutput(0);

    /// torch.cat *************************************************************************************************************
    ITensor* sum_input_tensors[] = {perm1, perm3};
    ILayer* y = network->addConcatenation(sum_input_tensors,2);

    Permutation permutation3 = {1,0,2};
    nvinfer1::IShuffleLayer* shuffleLayer3 = network->addShuffle(*y->getOutput(0));
    shuffleLayer3->setFirstTranspose(Permutation(permutation3));
    ITensor* perm4 = shuffleLayer3->getOutput(0);

    int mip = MAX(8, inp/32);

    /// conv1 *************************************************************************************************************
    auto y_conv = convBlock2(network, weightMap, *perm4, mip, 1, 1, 1, lname + ".conv1",lname);

    /// torch.split *************************************************************************************************************
    Dims d = y_conv->getOutput(0)->getDimensions();
    ISliceLayer* sliceLayer1 = network->addSlice(*y_conv->getOutput(0), Dims3(0, 0, 0), Dims3(d.d[0], d.d[1] / 2, d.d[2]), Dims3(1, 1, 1));
    ISliceLayer* sliceLayer2 = network->addSlice(*y_conv->getOutput(0), Dims3(0, d.d[1] / 2, 0), Dims3(d.d[0], d.d[1] / 2, d.d[2]), Dims3(1, 1, 1));

    Permutation permutation4 = {0,2,1};
    nvinfer1::IShuffleLayer* shuffleLayer4 = network->addShuffle(*sliceLayer2->getOutput(0));
    shuffleLayer4->setFirstTranspose(Permutation(permutation4));
    ITensor* perm5 = shuffleLayer4->getOutput(0);

    /// conv_h conv_w *************************************************************************************************************
    auto a_h = convBlock3(network, weightMap, *sliceLayer1->getOutput(0), inp, 1, 1, 1, lname + ".conv_h");
    auto a_w = convBlock3(network, weightMap, *perm5, inp, 1, 1, 1, lname + ".conv_w");

    ITensor* sum_input_tensors1[] = {&input, a_h->getOutput(0),a_w->getOutput(0)};
    auto ew1 = network->addElementWise(*sum_input_tensors1[0], *sum_input_tensors1[1], ElementWiseOperation::kPROD);
    auto ew2 = network->addElementWise(*ew1->getOutput(0), *sum_input_tensors1[2], ElementWiseOperation::kPROD);

    return ew2;
}
static ILayer* convBlock2(INetworkDefinition *network, std::map<std::string, Weights>& weightMap, ITensor& input,int outch, int ksize, int s,
                          int g, std::string lname1,std::string lname2)
{
    Weights emptywts{ DataType::kFLOAT, nullptr, 0 };
    int p = 0;

    IConvolutionLayer* conv1 = network->addConvolutionNd(input, outch, DimsHW{ ksize, ksize }, weightMap[lname1 + ".weight"], emptywts);
    assert(conv1);
    conv1->setStrideNd(DimsHW{ s, s });
    conv1->setPaddingNd(DimsHW{ p, p });
    conv1->setNbGroups(g);
    conv1->setName((lname1).c_str());

    IScaleLayer* bn1 = addBatchNorm2d(network, weightMap, *conv1->getOutput(0), lname2 + ".bn1", 1e-3);

    // silu = x * sigmoid
    auto sig = network->addActivation(*bn1->getOutput(0), ActivationType::kRELU);
    assert(sig);
    auto ew = network->addElementWise(*bn1->getOutput(0), *sig->getOutput(0), ElementWiseOperation::kPROD);
    assert(ew);
    return ew;
}

static ILayer* convBlock3(INetworkDefinition *network, std::map<std::string, Weights>& weightMap, ITensor& input, int outch, int ksize, int s, int g, std::string lname)
{
    Weights emptywts{ DataType::kFLOAT, nullptr, 0 };
    int p = 0;
    IConvolutionLayer* conv1 = network->addConvolutionNd(input, outch, DimsHW{ ksize, ksize }, weightMap[lname + ".weight"], emptywts);
    assert(conv1);
    conv1->setStrideNd(DimsHW{ s, s });
    conv1->setPaddingNd(DimsHW{ p, p });
    conv1->setNbGroups(g);
    conv1->setName((lname).c_str());

    // silu = x * sigmoid
    auto sig = network->addActivation(*conv1->getOutput(0), ActivationType::kSIGMOID);
    assert(sig);

    auto ew = network->addElementWise(*conv1->getOutput(0), *sig->getOutput(0), ElementWiseOperation::kPROD);

    return ew;
}

- what is your command? e.g. `sudo ./yolov5 -s`
- what's your output?
- what output do you expect?

**Comment by wang-xinyu** [comment]
You can try to debug it by getting the output of middle layers.

**Comment by wafaer** [comment]
> You can try to debug it by getting the output of middle layers.

Sorry, I'm a newbie, I don't quite understand what you mean, could you be more detailed, thank you.

**Comment by wafaer** [comment]
> For example, you have added 10 layers, from layer1 to layer10, now you only check the output tensor of layer10, and the result is wrong.
> 
> For debug purpose, you can first get the output of layer1, to see if the result is correct. If yes, then check layer2. Check the layer output one by one, until you found which layer has issue.

Hello, the modified model is the same as the original model. The Dimensions, DynamicRangeMin, DynamicRangeMax, and AllowedFormats of each layer's tensors are the same. What output should I check? If all the outputs are the same, what should I do?

**Comment by wang-xinyu** [comment]
Have you check the values of the output tensor for each layer?

**Comment by wafaer** [comment]
> Dimensions、DynamicRangeMin、DynamicRangeMax和AllowedFormats

Yes, I have checked every layer, including tensor Dimensions, DynamicRangeMin, DynamicRangeMax and AllowedFormats. They are all the same. Is it because I checked the tensor output type incorrectly? Let me reiterate my question: engine inference built after modifying the model. Is empty.

**Comment by wafaer** [comment]
> Have you check the values of the output tensor for each layer?

I don’t think we should check the Dimensions output by each layer’s tensor. If it is different from the source model, the engine model will not be generated, but I really don’t know what to check. Do you have any ideas? Specifically, we should check which outputs are wrong and lead to inferential reasoning.
